specification/compiler/steps/validate-model.ts

```diff
@@ -0,0 +1,520 @@
+        validateValueOf(v)
+      }
+      context.pop()
+    }
+  }
+
+  // -----------------------------------------------------------------------------------------------
+  // Type definitions validations
+
+  function validateTypeDef (typeDef: model.TypeDefinition, kind: TypeDefKind): void {
+    if (typesSeen.has(fqn(typeDef.name))) {
+      // Already visited
+      return
+    }
+    typesSeen.add(fqn(typeDef.name))
+
+    if (kind === TypeDefKind.request) {
+      if (typeDef.kind !== 'request') {
+        modelError(`Expected a request but type is a '${typeDef.kind}'`)
+      }
+    } else {
+      if (typeDef.kind === 'request') {
+        modelError('Type is a request, which should not be used for properties')
+      }
+    }
+
+    // Type definitions are top-level elements
+    const oldContext = context
+    context = []
+    context.push(`${typeDef.kind} definition ${typeDef.name.namespace}:${typeDef.name.name}`)
+
+    validateAnnotations(typeDef.annotations)
+
+    switch (typeDef.kind) {
+      case 'request':
+        validateRequest(typeDef)
+        break
+
+      case 'interface':
+        validateInterface(typeDef)
+        break
+
+      case 'enum':
+        validateEnum(typeDef)
+        break
+
+      case 'type_alias':
+        validateTypeAlias(typeDef)
+        break
+
+      case 'union':
+        validateUnion(typeDef)
+        break
+    }
+
+    // Restore previous context
+    context = oldContext
+  }
+
+  function validateRequest (typeDef: model.Request): void {
+    validateInherits(typeDef.inherits)
+    validateImplements(typeDef.implements)
+    validateBehaviors(typeDef)
+
+    context.push('path')
+    validateProperties(typeDef.path)
+    context.pop()
+
+    context.push('query')
+    validateProperties(typeDef.query)
+    context.pop()
+
+    if (typeDef.body != null) {
+      context.push('body')
+      switch (typeDef.body.kind) {
+        case 'properties':
+          validateProperties(typeDef.body.properties)
+          break
+        case 'value':
+          validateValueOf(typeDef.body.value)
+          break
+      }
+      context.pop()
+    }
+  }
+
+  function validateInterface (typeDef: model.Interface): void {
+    validateImplements(typeDef.implements)
+    validateInherits(typeDef.inherits)
+    validateBehaviors(typeDef)
+    validateProperties(typeDef.properties)
+  }
+
+  function validateEnum (typeDef: model.Enum): void {
+    validateAnnotations(typeDef.annotations)
+    const values = new Set<string>()
+    for (const member of typeDef.members) {
+      const lcName = member.name.toLowerCase()
+      if (values.has(lcName)) {
+        modelError(`Duplicate lowercase enum name: ${member.name}`)
+      }
+      values.add(lcName)
+    }
+  }
+
+  function validateTypeAlias (typeDef: model.TypeAlias): void {
+    validateAnnotations(typeDef.annotations)
+    // FIXME: do we really need generics on a type alias? An alias is semantically similar to a 1-member union
```

Comment:
We do, this came through as part of: https://github.com/elastic/elastic-client-generator/pull/151/files

This means generics from types can be passed down to type aliases on properties without losing content. 


Union itself can not be generic but the items it contain can